File: 0778-swim-in-rising-water/0778-swim-in-rising-water.py
class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        #Dijikstra's + BFS

        rows = len(grid)
        cols = len(grid[0])

        minHeap = []
        minHeap.append((grid[0][0], 0, 0)) #w, x, y
        visited = set()
        directions = [(0, -1), (0, 1), (-1, 0), (1, 0)] #Up, Down, Left, Right
        
        while minHeap:
            w, x, y = heapq.heappop(minHeap)

            if (x, y) in visited:
                continue
            visited.add((x, y))

            if x == rows - 1 and y == cols - 1:
                if grid[x][y] > w:
                    return grid[x][y]
                else:
                    return w

            for nr, nc in directions:
                newRow = x + nr
                newCol = y + nc

                #Bounds
                if 0 <= newRow < rows and 0 <= newCol < cols and (newRow, newCol) not in visited:
                    if grid[newRow][newCol] > w:
                        heapq.heappush(minHeap, (grid[newRow][newCol], newRow, newCol))
                    else:
                        heapq.heappush(minHeap, (w, newRow, newCol))




        

File: 0778-swim-in-rising-water/0778-swim-in-rising-water.cpp
class Solution {
public:
    int swimInWater(vector<vector<int>>& grid) {
        int n = grid.size();
        vector<vector<int>> T(n, vector<int>(n, INT_MAX)); 
        vector<pair<int,int>> directions = {{0,1},{1,0},{0,-1},{-1,0}};

        queue<pair<int,int>> q;
        T[0][0] = grid[0][0];
        q.push({0,0});

        while(!q.empty()) {
            auto [r,c] = q.front(); q.pop();

            for(auto &[dr, dc] : directions) {
                int nr = r + dr, nc = c + dc;

                if(nr >= 0 && nr < n && nc >= 0 && nc < n) {
                    int newT = max(T[r][c], grid[nr][nc]);
                    if(newT < T[nr][nc]) {
                        T[nr][nc] = newT;
                        q.push({nr,nc});
                    }
                }
            }
        }

        return T[n-1][n-1];
    }
};